### DISCUSSION TITLE
Maybe document when context manager shouldn't be used

### DISCUSSION THREAD

**Discussion by trondhindenes** [encode/httpx#2444] [Q&A] [answered]
Quite recently we started using FastAPI to develop small async-capable apis, and whenever we need to invoke http requests from an async endpoint we use httpx.

For a while we just followed the documentation and dropped
```
>>> async with httpx.AsyncClient() as client:
...     r = await client.get('https://www.example.com/')
```
In our code. However, in some of our apis we saw noticeable slowdown/excessive cpu usage and we (after much troubleshooting) figured out that this was a service that performs a _lot_ of http requests to another service. From there we learned that instantiating httpx using a context manager as shown above has significant performance implications.

After learning that, we started instatiating a "global" httpx client, and used that throughout the code instead. The results were night and day.

I'm wondering if the documentation should be more nuanced around this, and maybe disuss situations where the context manager shouldn't be used but one should instead instantiate a more long-lived client and reuse that.

**Comment by lovelydinosaur** [comment, marked as answer]
Hi @trondhindenes.

I've attempted to give an answer to this here... https://github.com/encode/httpx/discussions/2191

I'm a bit uncertain of how we ought to update the documentation. I'd be happier if we were able to recommend *lifespan* based clients, rather than *global* clients, *particularly in the async case*, but. 🤷‍♂️

**Reply by Kludex** [reply]
cc @adriangb /\ :eyes:

**Reply by trondhindenes** [reply]
Thanks, that makes sense.
FastAPI actually has startup and teardown hooks that probably fit well here, but I guess we're looking for a more global (har har) recommendation.

I'm just bringing this up because we spent a long time understanding what was happening, since we ported a service to async and modern stuff we couldn't believe how bad the performance was. Might help someone if the docs mention that "you shouldn't always use the context manager pattern" more clearly

**Reply by adriangb** [reply]
I think the docs could say something like “creating an AsyncClient is expensive. If you are doing this in a hot loop you should find a way to hoist the client outside of the loop (for example using startup/shutdown events in ASGI applications)”

**Reply by trondhindenes** [reply]
Perfect

**Reply by adriangb** [reply]
https://github.com/encode/httpx/pull/2452
